tensorboard/plugins/projector/vz_projector/data.ts

```diff
@@ -322,20 +324,25 @@ export class DataSet {
         this.tsne = null;
         return;
       }
-      this.tsne.step();
-      let result = this.tsne.getSolution();
-      sampledIndices.forEach((index, i) => {
-        let dataPoint = this.points[index];
-
-        dataPoint.projections['tsne-0'] = result[i * tsneDim + 0];
-        dataPoint.projections['tsne-1'] = result[i * tsneDim + 1];
-        if (tsneDim === 3) {
-          dataPoint.projections['tsne-2'] = result[i * tsneDim + 2];
-        }
-      });
-      this.tSNEIteration++;
-      stepCallback(this.tSNEIteration);
-      requestAnimationFrame(step);
+      if (!this.tSNEShouldPause) {
+        this.tsne.step();
+        let result = this.tsne.getSolution();
+        sampledIndices.forEach((index, i) => {
+          let dataPoint = this.points[index];
+
+          dataPoint.projections['tsne-0'] = result[i * tsneDim + 0];
+          dataPoint.projections['tsne-1'] = result[i * tsneDim + 1];
+          if (tsneDim === 3) {
+            dataPoint.projections['tsne-2'] = result[i * tsneDim + 2];
+          }
+        });
+        this.tSNEIteration++;
+        stepCallback(this.tSNEIteration);
+        requestAnimationFrame(step);
+      }
+      else {
+        setTimeout(function (){requestAnimationFrame(step); }, 100);
```

Comment:
why the combined setTimeout & rAF?

this can be simplified, either way:

setTimeout(() => requestAnimationFrame(step), 100);

Reply:
Seems I should use rAF directly yes, thanks. I'm reconsidering and simplifying the changes to integrate pause-resume into the Stop button. I'll split the pause-resume and sprite zoom functionalities and supersede with new pull requests.